pbft/pbft.cpp

```diff
@@ -369,6 +372,51 @@ pbft::handle_join_or_leave(const pbft_membership_msg& msg)
     }
 }
 
+void
+pbft::handle_get_state(const pbft_membership_msg& msg, std::shared_ptr<bzn::session_base> session) const
+{
+    // get stable checkpoint for request
+    checkpoint_t req_cp(msg.sequence(), msg.state_hash());
+
+    if (req_cp == this->latest_stable_checkpoint())
+    {
+        pbft_membership_msg reply;
+        reply.set_type(PBFT_MMSG_SET_STATE);
+        reply.set_sequence(req_cp.first);
+        reply.set_state_hash(req_cp.second);
+        reply.set_state_data(this->get_checkpoint_state(req_cp));
+
+        auto msg_ptr = std::make_shared<bzn::encoded_message>(this->wrap_message(reply));
+        session->send_datagram(msg_ptr);
+    }
+    else
+    {
+        LOG(debug) << boost::format("Request for checkpoint that I don't have: seq: %1%, hash: %2%")
+            % msg.sequence(), msg.state_hash();
+    }
+}
+
+void
+pbft::handle_set_state(const pbft_membership_msg& msg)
+{
+    checkpoint_t cp(msg.sequence(), msg.state_hash());
+
+    if (this->unstable_checkpoint_proofs[cp].size() >= this->quorum_size() &&
```

Comment:
Can you put an explanatory comment on this check? It took me several takes to understand its intent as "if we have checkpoint messages, but not the state, for the checkpoint"